File: examples/sequential-cpp/example3.cpp
/* This code calculates the sum of all elements in an array */
#include <iostream>
#include <cmath>
#include "utils/cppheader.h"

const int SIZE = 1000000000;

using namespace std;

class Deviation {
private:
	int *myArray, mySize;
	double result;
	
public:
	Deviation(int *array, int size) 
		: myArray(array), mySize(size), result(0) {}
	
	double getResult() const {
		return result;
	}
	
	void calculate() {
		int i;
		double acum, avg;
		
		acum = 0;
		for (i = 0; i < mySize; i++) {
			acum += myArray[i];
		}
		avg = acum / mySize;
		
		acum = 0;
		for (i = 0; i < mySize; i++) {
			acum += (myArray[i] - avg) * (myArray[i] - avg);
		}
		result = sqrt(acum / mySize);
	}
};

int main(int argc, char* argv[]) {
	double ms;
	Timer t;
	int *a;
	
	a = new int[SIZE];
	random_array(a, SIZE);
	display_array("a", a);
	
	
	ms = 0;
	Deviation obj(a, SIZE);
	cout << "Starting..." << endl;
	for (int i = 0; i < N; i++) {
		t.start();
		obj.calculate();
		ms += t.stop();
	}
	cout << "S = " << obj.getResult() << endl;
	cout << "avg time = " << (ms / N) << " ms" << endl;
	
	delete [] a;
	return 0;
}

File: examples/sequential-java/Example3.java
/* This code adds all the values of an array */

public class Example3 {
	private int array[];
	private double result;
	
	public Example3(int array[]) {
		this.array = array;
		this.result = 0;
	}
	
	public double getResult() {
		return result;
	}
	
	public void calculate() {
		double acum, avg;
		int i;
	
		acum = 0;
		for (i = 0; i < array.length; i++) {
			acum += array[i];
		}
		avg = acum / array.length;
		
		acum = 0;
		for (i = 0; i < array.length; i++) {
			acum += (array[i] - avg) * (array[i] - avg);
		}
		
		result = Math.sqrt(acum / array.length);
	}
	
	public static void main(String args[]) {
		final int SIZE = 300_000_000;
		int array[] = new int[SIZE];
		long startTime, stopTime;
		double acum = 0;
		
		Utils.randomArray(array);
		Utils.displayArray("array", array);
		
		Example3 e = new Example3(array);
		acum = 0;
		for (int i = 0; i < Utils.N; i++) {
			startTime = System.currentTimeMillis();
			e.calculate();
			stopTime = System.currentTimeMillis();
			acum += (stopTime - startTime);
		}
		System.out.printf("S = %d\n", e.getResult());
		System.out.printf("avg time = %.5f\n", (acum / Utils.N));
	}
}
			
